// passport-youtube-v3/lib/passport-youtube-v3/strategy.js
/**
 * Module dependencies.
 */
var util = require('util')
  , OAuth2Strategy = require('passport-oauth').OAuth2Strategy
  , InternalOAuthError = require('passport-oauth').InternalOAuthError
  , parse = require('url-parse');  

//Información del usuario
var channels = {};
var channelSections = {};
var activities = {};
var commentThreads = {};
var allCommentThreads = {};
var playListItems = {};
var videos = {};

/**
 * `Strategy` constructor.
 *
 * Youtube authentication strategy authenticates requests using the OAuth 2.0 protocol.
 *
 * @param {Object} options
 * @param {Function} verify
 * @api public
 */
function Strategy(options, verify) {
  options = options || {};
  options.authorizationURL = options.authorizationURL || 'https://accounts.google.com/o/oauth2/auth';
  options.tokenURL = options.tokenURL || 'https://accounts.google.com/o/oauth2/token';
  options.scope = options.scope || ['https://www.googleapis.com/auth/youtube'];

  OAuth2Strategy.call(this, options, verify);
  this.name = 'youtube';
  this._profileURL = options.profileURL || 'https://www.googleapis.com/youtube/v3/channels?part=snippet&mine=true';
  this._authorizationParams = options.authorizationParams;
}

/**
 * Inherit from `OAuth2Strategy`.
 */
util.inherits(Strategy, OAuth2Strategy);

/**
 * Retrieve user profile from Youtube.
 *
 * This function constructs a normalized profile, with the following properties:
 *
 *   - `provider`         always set to `youtube`
 *   - `id`               the user's Google Plus user ID
 *   - `username`         the user's Youtube username
 *   - `displayName`      the user's full name
 *   - `name.familyName`  the user's last name
 *   - `name.givenName`   the user's first name
 *
 * @param {String} accessToken
 * @param {Function} done
 * @api protected
 */

      /*
      if (youtubeProfile) {
        profile.id = youtubeProfile.id;
        profile.displayName = youtubeProfile.snippet.title;
      }
	done(null, profile);
*/
      //Exportamos los metdos get() para obtener las variables con los datos guardados
      module.exports.getChannel = getChannel;
      module.exports.getChannelSections = getChannelSections;
      module.exports.getActivities = getActivities;
      module.exports.getCommentThreads = getCommentThreads;
      module.exports.getAllCommentThreads = getAllCommentThreads;
      module.exports.getPlayListItems = getPlayListItems;
      module.exports.getVideos = getVideos;

Strategy.prototype.userProfile = function(accessToken, done) {
  var channel_id;
  var playlistId;
  //
  var url_channelSection = 'https://www.googleapis.com/youtube/v3/channelSections?part=contentDetails&mine=true';
  var url_channels = 'https://www.googleapis.com/youtube/v3/channels?part=contentDetails%2C+snippet%2C+status%2C+statistics%2C+contentOwnerDetails+&maxResults=50&mine=true';
  //
  var url_activities = 'https://www.googleapis.com/youtube/v3/activities?part=contentDetails%2C+snippet&maxResults=50&mine=true';
  //requieren channel_id del cliente
  var url_temp;
  var url_allCommentThreads; 
  var url_commentThreads;
  //CAMBIAR ID's
  //Los ID de playlists y videos se pueden obtener de la cuenta del canal, al igual que los comentarios
  //Se necesita cambiar el ID dependiendo que tipo de lista queremos obtener (historial, likes, playlist), para prueba usamos playListId=WLyvjBEjZyVP19xUv4T8o0pw
  var url_playListItems = 'https://www.googleapis.com/youtube/v3/playlistItems?part=snippet%2C+contentDetails%2C+status&maxResults=50&playlistId=WLyvjBEjZyVP19xUv4T8o0pw';
  //Se requiere el iD de un video en especifico que se pueden obtener de las playslists
  var url_videos= 'https://www.googleapis.com/youtube/v3/videos?part=contentDetails%2C+status%2C+statistics%2C+snippet&id=8I5g5pj0tMU';
  var profile = {};
  try {
   this._oauth2.getProtectedResource(url_channels, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      channels = json;
      channel_id = channels.items[0].id;
      //asigna el ID del canal a las url's
      url_temp = 'https://www.googleapis.com/youtube/v3/commentThreads?part=snippet&allThreadsRelatedToChannelId='+channel_id+'&maxResults=1&order=time&textFormat=plainText';
      url_allCommentThreads = parse(url_temp, true);
      url_temp = 'https://www.googleapis.com/youtube/v3/commentThreads?part=snippet&channelId='+channel_id+'&maxResults=1&order=time&textFormat=plainText';
      url_commentThreads = parse(url_temp, true);
  });
   this._oauth2.getProtectedResource(url_channelSection, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      channelSections = json;
  });
  this._oauth2.getProtectedResource(url_activities, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      activities = json;
  });
  this._oauth2.getProtectedResource(url_commentThreads, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      commentThreads= json;
  });  
  this._oauth2.getProtectedResource(url_allCommentThreads, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      allCommentThreads = json;
  });  
  this._oauth2.getProtectedResource(url_playListItems, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      playListItems = json;
  });
  this._oauth2.getProtectedResource(url_videos, accessToken, function (err, body, res) {
      if (err) { return done(new InternalOAuthError('failed to fetch user profile', err)); }
      var json = JSON.parse(body);
      videos = json;
      done(null, profile);
  }); 
    } catch(e) {
      done(e);
    }
    
}


Strategy.prototype._convertProfileFields = function(profileFields) {
  var map = {
    'id':          'id',
    'username':    'username',
    'displayName': 'name',
    'name':       ['last_name', 'first_name']
  };

  var fields = [];

  profileFields.forEach(function(f) {
    if (typeof map[f] === 'undefined') return;

    if (Array.isArray(map[f])) {
      Array.prototype.push.apply(fields, map[f]);
    } else {
      fields.push(map[f]);
    }
  });

  return fields.join(',');
}


Strategy.prototype.authorizationParams = function(options) {
  return options && options.authorizationParams || this._authorizationParams || {
    access_type : 'online',
    approval_prompt: 'auto'
  }
}

var getChannel = function() {
  return channels;
}
var getChannelSections = function() {
  return channelSections;
}
var getActivities = function() {
  return activities;
}
var getCommentThreads = function() {
  return commentThreads;
}
var getAllCommentThreads = function() {
  return allCommentThreads;
}
var getPlayListItems = function () {
  return playListItems;
}
var getVideos = function () {
  return videos;
}
/**
 * Expose `Strategy`.
 */

module.exports = Strategy;
